pallets/services/src/lib.rs

```diff
@@ -1099,5 +1054,59 @@ pub mod module {
 			// TODO: add weight for the call to the total weight.
 			Ok(PostDispatchInfo { actual_weight: None, pays_fee: Pays::Yes })
 		}
+
+		pub fn slash(
+			origin: OriginFor<T>,
+			offender: T::AccountId,
+			#[pallet::compact] service_id: u64,
+			#[pallet::compact] percent: Percent,
+		) -> DispatchResult {
+			let caller = ensure_signed(origin)?;
+			let service = Self::services(service_id)?;
+			// TODO: add EVM hook to query for the slashing permission.
+			ensure!(service.owner == caller, DispatchError::BadOrigin);
+			let service_operator =
+				service.operators.iter().find(|(operator, _)| operator == &offender);
+			let (operator, restake_percent) = match service_operator {
+				Some((operator, restake_percent)) => (operator, restake_percent),
+				None => return Err(Error::<T>::OffenderNotOperator.into()),
+			};
+			// sanity check
+			ensure!(operator == &offender, Error::<T>::OffenderNotOperator);
+
+			let operator_is_active = T::OperatorDelegationManager::is_operator_active(&offender);
+			ensure!(operator_is_active, Error::<T>::OffenderNotActiveOperator);
+
+			let total_own_stake = T::OperatorDelegationManager::get_operator_stake(&offender);
+			// Only take the exposed restake percentage for this service.
+			let own_stake = restake_percent.mul_floor(total_own_stake);
+			let delegators = T::OperatorDelegationManager::get_delegators_for_operator(&offender);
+			let own_slash = percent.mul_floor(own_stake);
+			let others_slash = delegators
+				.into_iter()
+				.map(|(delegator, stake, _asset_id)| (delegator, percent.mul_floor(stake)))
```

Comment:
Why remove knowledge of the asset when you return it? Isn't this necessary across all assets being used by the service?

Reply:
we currently treat them equally, i.e if the slash is 10%, then we slash every delegator's stake by the 10% regardless of the asset kind. this not ideal of course, since not all assets are equal. Until we have an interface to map everything to Native token (TNT) we can take asset kind into account.